Signal/src/ViewControllers/OWS2FASettingsViewController.m

```diff
@@ -0,0 +1,493 @@
+{
+    // Note: This affects how the "back" button will look if another
+    //       view is pushed on top of this one, not how the "back"
+    //       button looks when this view is visible.
+    self.navigationItem.backBarButtonItem =
+        [[UIBarButtonItem alloc] initWithTitle:NSLocalizedString(@"VERIFICATION_BACK_BUTTON",
+                                                   @"button text for back button on verification view")
+                                         style:UIBarButtonItemStylePlain
+                                        target:self
+                                        action:@selector(backButtonWasPressed)];
+
+    if (self.shouldHaveNextButton) {
+        self.navigationItem.rightBarButtonItem = [[UIBarButtonItem alloc]
+            initWithTitle:NSLocalizedString(@"ENABLE_2FA_VIEW_NEXT_BUTTON",
+                              @"Label for the 'next' button in the 'enable two factor auth' views.")
+                    style:UIBarButtonItemStylePlain
+                   target:self
+                   action:@selector(nextButtonWasPressed)];
+    } else {
+        self.navigationItem.rightBarButtonItem = nil;
+    }
+}
+
+#pragma mark - UITextFieldDelegate
+
+- (BOOL)textField:(UITextField *)textField
+    shouldChangeCharactersInRange:(NSRange)range
+                replacementString:(NSString *)insertionText
+{
+    // TODO: ?
+    const NSUInteger kMaxPinLength = 14;
+
+    // * We only want to let the user enter decimal digits.
+    // * The user should be able to copy and paste freely.
+    // * Invalid input should be simply ignored.
+    //
+    // We accomplish this by being permissive and trying to "take as much of the user
+    // input as possible".
+    //
+    // * Always accept deletes.
+    // * Ignore invalid input.
+    // * Take partial input if possible.
+
+    NSString *oldText = textField.text;
+    // Construct the new contents of the text field by:
+    // 1. Determining the "left" substring: the contents of the old text _before_ the deletion range.
+    //    Filtering will remove non-decimal digit characters.
+    NSString *left = [oldText substringToIndex:range.location].digitsOnly;
+    // 2. Determining the "right" substring: the contents of the old text _after_ the deletion range.
+    NSString *right = [oldText substringFromIndex:range.location + range.length].digitsOnly;
+    // 3. Determining the "center" substring: the contents of the new insertion text.
+    NSString *center = insertionText.digitsOnly;
+    // 4. Construct the "raw" new text by concatenating left, center and right.
+    NSString *textAfterChange = [[left stringByAppendingString:center] stringByAppendingString:right];
+    // 5. Ensure we don't exceed the maximum length for a PIN.
+    if (textAfterChange.length > kMaxPinLength) {
+        textAfterChange = [textAfterChange substringToIndex:kMaxPinLength];
+    }
+    // 6. Construct the "formatted" new text by inserting a hyphen if necessary.
```

Comment:
I don't see where we're inserting a hyphen. Is this just an old copy/paste comment?

Reply:
Done.